pkg/handlers/generic/lifecycle/ccm/handler.go

```diff
@@ -67,14 +68,39 @@ func (c *CCMHandler) AfterControlPlaneInitialized(
 	req *runtimehooksv1.AfterControlPlaneInitializedRequest,
 	resp *runtimehooksv1.AfterControlPlaneInitializedResponse,
 ) {
-	clusterKey := ctrlclient.ObjectKeyFromObject(&req.Cluster)
+	commonResponse := &runtimehooksv1.CommonResponse{}
+	c.apply(ctx, &req.Cluster, commonResponse)
+	resp.Status = commonResponse.GetStatus()
+	resp.Message = commonResponse.GetMessage()
+}
+
+func (c *CCMHandler) BeforeClusterUpgrade(
+	ctx context.Context,
+	req *runtimehooksv1.BeforeClusterUpgradeRequest,
+	resp *runtimehooksv1.BeforeClusterUpgradeResponse,
+) {
+	commonResponse := &runtimehooksv1.CommonResponse{}
+	c.apply(ctx, &req.Cluster, commonResponse)
+	resp.Status = commonResponse.GetStatus()
+	resp.Message = commonResponse.GetMessage()
+	if resp.Status == runtimehooksv1.ResponseStatusFailure {
+		resp.SetRetryAfterSeconds(lifecycle.BeforeClusterUpgradeRetryAfterSeconds)
+	}
```

Comment:
Retry can be set even with success response, and should be set in the case of waiting for addons to actually be successfully deployed and running. I think we'll need a different signature for apply, something like:

```
func (c *CCMHandler) apply(
	ctx context.Context,
	cluster *clusterv1.Cluster,
	resp *runtimehooksv1.CommonResponse,
) (completed bool, err error) // No need for named returns, just here for clarity
```

The `BeforeClusterUpgrade` hook can set the retry time if completed is false, and `AfterInitializedControlPlane` can ignore `completed`.

In the `apply` function, the logic will have to check for the existence and status of the applied resources and preferably the status of the actual deployed resources in workload cluster to make the decision about whether it has `completed` or not.

Reply:
That sounds great to me. 

thanks for the explanation.